yesod-form/Yesod/Form/Fields.hs

```diff
@@ -427,25 +428,39 @@ selectFieldList :: (Eq a, RenderMessage site FormMessage, RenderMessage site msg
                 -> Field (HandlerFor site) a
 selectFieldList = selectField . optionsPairs
 
+-- | Creates a @\<select>@ tag with @\<optgroup>@s for selecting one option.
+selectFieldListGrouped :: (Eq a, RenderMessage site FormMessage, RenderMessage site msg)
+                => [(msg, [(msg, a)])]
+                -> Field (HandlerFor site) a
+selectFieldListGrouped = selectField . optionsPairsGrouped
+
 -- | Creates a @\<select>@ tag for selecting one option. Example usage:
 --
 -- > areq (selectField $ optionsPairs [(MsgValue1, "value1"),(MsgValue2, "value2")]) "Which value?" Nothing
 selectField :: (Eq a, RenderMessage site FormMessage)
             => HandlerFor site (OptionList a)
             -> Field (HandlerFor site) a
-selectField = selectFieldHelper
-    (\theId name attrs inside -> [whamlet|
+selectField ol = do
+  -- ol' <- ol -- TODO how to extract the option list for pattern matching?
```

Comment:
@snoyberg I need help

Reply:
Please look at the implementation of `selectFieldHelper`. It looks like you'll need to modify things in there. Overall it seems like the approach you're trying to take is incompatible with how these functions are structured.

Reply:
... because `selectField` is a pure function and I can pass but not modify monadic values... I think I understand now.